query/src/main/java/org/infinispan/query/backend/QueryInterceptor.java

```diff
@@ -322,69 +314,66 @@ public Object visitReadWriteManyEntriesCommand(InvocationContext ctx, ReadWriteM
     * Remove all entries from all known indexes
     */
    public void purgeAllIndexes() {
-      purgeAllIndexes(NoTransactionContext.INSTANCE);
+      SearchMapping searchMapping = searchMappingHolder.getSearchMapping();
+      if (searchMapping == null) {
+         return;
+      }
+
+      searchMapping.scopeAll().workspace().purge();
    }
 
    public void purgeIndex(Class<?> entityType) {
-      purgeIndex(NoTransactionContext.INSTANCE, entityType);
+      SearchMapping searchMapping = searchMappingHolder.getSearchMapping();
+      if (searchMapping == null) {
+         return;
+      }
+
+      searchMapping.scope(entityType).workspace().purge();
+   }
+
+   public void refreshIndex(Class<?> entityType) {
+      SearchMapping searchMapping = searchMappingHolder.getSearchMapping();
+      if (searchMapping == null) {
+         return;
+      }
+
+      searchMapping.scope(entityType).workspace().refresh();
    }
 
    /**
     * Removes from the index the entries corresponding to the supplied segments, if the index is local.
     */
    void purgeIndex(IntSet segments) {
-      if (segments == null) return;
-      for (int segment : segments) {
-         DeletionQuery deletionQuery = new SingularTermDeletionQuery(SEGMENT_FIELD, String.valueOf(segment));
-         for (IndexedTypeIdentifier type : searchFactory.getIndexBindings().keySet()) {
-            Work deleteWork = new DeleteByQueryWork(type, deletionQuery);
-            performSearchWork(deleteWork, NoTransactionContext.INSTANCE);
-         }
+      if (segments == null || segments.isEmpty()) return;
+
+      SearchMapping searchMapping = searchMappingHolder.getSearchMapping();
+      if (searchMapping == null) {
```

Comment:
There was for the same reason of https://github.com/infinispan/infinispan/pull/8571#discussion_r458793497.
Going to remove these checks...
Moreover, probably thanks to https://github.com/infinispan/infinispan/pull/8575 we can get rid of the entire `SearchMappingHolder` class, but maybe this could be a follow up. What do you think about that? Thanks

Reply:
Looking deeper at the issue, I think that we should keep these null checks or change the way we handle the case where we declare a cache as indexed but we don't provide any entity in the `indexedEntities` configuration.

Basically, if we remove these checks, we will have these three tests not working:
1. `ContinuousQueryObjectStorageTest`
2. `DataFormatIndexedTest`
3. `ClientListenerWithDslFilterObjectStorageTest`

In all these cases we configure the cache in this way:
``` java
cfgBuilder.indexing().enable()
            .addProperty("directory.type", "local-heap");
```

The cache is indexed but we don’t provide any indexed entity in the configuration.
I think that it would be the same if we defined some entities, but all of them were not contained in the serialization context at the time of cache starting.

Without the check we will get a null pointer exception locally, that will produce a timeout exception remotely.
The is issue is bound to the methods:
1. QueryInterceptor#purgeAllIndexes
2. QueryInterceptor#purgeIndex by segment

To be honest, I have no idea why a purge could be invoked on an indexed cache without no entities. Can you help me with that? Thanks

Reply:
Let's just handle it in a follow up

Reply:
Let's assume ```indexedEntities``` will always have at least 1 declared entity starting with ispn 12